### PR TITLE
[BE-24] Create Administrator package with entity, service, controller…

### PR DESCRIPTION
**Pull request by nidaffs**
…, repository, dto

### CODE REVIEW

File: src/main/java/com/exadel/project/administrator/service/AdministratorService.java [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+@RequiredArgsConstructor
+public class AdministratorService extends BaseService<Administrator, AdministratorRepository> {
+
+    private final AdministratorMapper administratorMapper;
+
+    @Autowired
+    private PasswordEncoder passwordEncoder;
+
+    @Autowired
+    private AdministratorRepository administratorRepository;
```
Comment (serkushner) [inline src/main/java/com/exadel/project/administrator/service/AdministratorService.java:29]: Pls, make constructor injection here

File: src/main/java/com/exadel/project/administrator/service/AdministratorService.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        Administrator administrator = new Administrator();
+        administrator.setLogin(administratorDto.getLogin());
+        administrator.setEmail(administratorDto.getEmail());
+        administrator.setName(administratorDto.getName());
+        administrator.setSurname(administratorDto.getSurname());
+        administrator.setPassword(passwordEncoder.encode(administratorDto.getPassword()));
+        administrator.setRole(Role.ADMIN);
+        administrator.setPhone(administratorDto.getPhone());
+        administrator.setSkype(administratorDto.getSkype());
```
Comment (serkushner) [inline src/main/java/com/exadel/project/administrator/service/AdministratorService.java:52]: Lets move this logic into mapper, so your service should concentrate on business logic

File: src/main/java/com/exadel/project/administrator/service/AdministratorService.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+        }
+        if (StringUtils.isNotBlank(administratorDto.getPassword())) {
+            administrator.setPassword(passwordEncoder.encode(administratorDto.getPassword()));
+        }
+        if (StringUtils.isNotBlank(administratorDto.getPhone())) {
+            administrator.setPhone(administratorDto.getPhone());
+        }
+        if (StringUtils.isNotBlank(administratorDto.getSkype())) {
+            administrator.setSkype(administratorDto.getSkype());
+        }
```
Comment (serkushner) [inline src/main/java/com/exadel/project/administrator/service/AdministratorService.java:78]: pls lets do all mapping logic into mapper component

File: src/main/resources/db/changelog/changelog-init-tables.yaml [hunk 4]
```diff
@@ -158,7 +158,7 @@ databaseChangeLog:
                     nullable: false
               - column:
                   name: password
-                  type: VARCHAR(30)
+                  type: VARCHAR(255)
```
Comment (serkushner) [inline src/main/resources/db/changelog/changelog-init-tables.yaml:161]: Pls, lets do this change in separate file

**Review by serkushner — CHANGES_REQUESTED** [review]

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+import com.exadel.project.administrator.service.AdministratorService;
+import com.exadel.project.common.exception.EntityAlreadyExistsException;
+import com.exadel.project.common.exception.EntityNotFoundException;
+import lombok.RequiredArgsConstructor;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.http.MediaType;
+import org.springframework.web.bind.annotation.*;
+
+@RestController
+@RequestMapping(value = "/admin/")
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:14]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+public class AdministratorController {
+
+    @Autowired
+    private AdministratorService administratorService;
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:18]: thanks, fixed

File: src/main/resources/db/changelog/changelog-init-tables.yaml (hunk 4 above)
Comment (nidaffs) [reply src/main/resources/db/changelog/changelog-init-tables.yaml:161]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+        try {
+            return administratorService.getAdministratorById(id);
+        } catch (EntityNotFoundException e) {
+            System.out.println(e.getMessage());
+            return null;
+        }
+    }
+
+    @GetMapping(value = ID +"/delete")
+    public boolean deleteAdmin(@PathVariable Long id) {
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:33]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+            administratorService.deleteAdministratorById(id);
+            return true;
+        } catch (EntityNotFoundException e) {
+            System.out.println(e.getMessage());
+            return false;
+        }
+    }
+
+    @PostMapping(value = "add", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    public Administrator addAdministrator(@RequestBody AdministratorDto dto) {
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:44]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 9]
```diff
@@ -0,0 +1,71 @@
...
+public class AdministratorController {
+
+    @Autowired
+    private AdministratorService administratorService;
+
+    private static final String ID = "/{id}";
+
+    @GetMapping(value = ID)
+    public AdministratorDto getAdmin(@PathVariable Long id) {
+        try {
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:24]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 10]
```diff
@@ -0,0 +1,71 @@
...
+    public Administrator addAdministrator(@RequestBody AdministratorDto dto) {
+        try {
+            return administratorService.addAdministrator(dto);
+        } catch (EntityAlreadyExistsException e) {
+            System.out.println(e.getMessage());
+            return null;
+        }
+    }
+
+    @PostMapping(value = ID +"/update", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:53]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+        try {
+            administratorService.deleteAdministratorById(id);
+            return true;
+        } catch (EntityNotFoundException e) {
+            System.out.println(e.getMessage());
+            return false;
+        }
+    }
+
+    @PostMapping(value = "add", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:43]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 12]
```diff
@@ -0,0 +1,71 @@
...
+        try {
+            return administratorService.addAdministrator(dto);
+        } catch (EntityAlreadyExistsException e) {
+            System.out.println(e.getMessage());
+            return null;
+        }
+    }
+
+    @PostMapping(value = ID +"/update", consumes = MediaType.APPLICATION_JSON_VALUE, produces = MediaType.APPLICATION_JSON_VALUE)
+    public void updateAdministrator(@PathVariable Long id, @RequestBody AdministratorDto dto) {
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:54]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 13]
```diff
@@ -0,0 +1,71 @@
...
+        try {
+            administratorService.updateAdministrator(id, dto);
+        } catch (EntityNotFoundException e) {
+            System.out.println(e.getMessage());
+        }
+    }
+
+    @PostMapping(value = ID +"/changerole")
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:62]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/controller/AdministratorController.java [hunk 14]
```diff
@@ -0,0 +1,71 @@
...
+    public void changeAdministratorRole(@PathVariable Long id, @RequestBody String roleName) {
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/controller/AdministratorController.java:63]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/service/AdministratorService.java (hunk 1 above)
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/service/AdministratorService.java:29]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/service/AdministratorService.java (hunk 2 above)
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/service/AdministratorService.java:52]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/service/AdministratorService.java (hunk 3 above)
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/service/AdministratorService.java:78]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/entity/Administrator.java [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+    @NotBlank(message = "Administrator surname should not be empty")
+    @Column(name = "surname")
+    private String surname;
+
+    @NotBlank(message = "Administrator phone should not be empty")
+    @Column(name = "phone")
+    private String phone;
+
+    @NotBlank(message = "Administrator skype should not be empty")
+    @Column(name = "skype")
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/entity/Administrator.java:45]: thanks, fixed

File: src/main/java/com/exadel/project/administrator/entity/Administrator.java [hunk 16]
```diff
@@ -0,0 +1,52 @@
...
+    @Column(name = "id")
+    private Long id;
+
+    @NotBlank(message = "Administrator login should not be empty")
+    @Column(name = "login")
+    private String login;
+
+    @NotBlank(message = "Administrator email should not be empty")
+    @Column(name = "email")
+    private String email;
```
[the remark this answers is not in the document]
Comment (nidaffs) [reply src/main/java/com/exadel/project/administrator/entity/Administrator.java:26]: thanks, fixed

**Review by serkushner — APPROVED** [review]

**Review by AlKaliada — APPROVED** [review]

**Pull request merged**